enterprise/tailnet/pgcoord.go

```diff
@@ -0,0 +1,1207 @@
+		}
+	}
+	q.mu.Lock()
+	mpr, ok := q.mappers[mk]
+	q.mu.Unlock()
+	if !ok {
+		q.logger.Debug(q.ctx, "query for missing mapper",
+			slog.F("agent_id", mk.agent), slog.F("clients_of_agent", mk.clientsOfAgent))
+		return nil
+	}
+	mpr.mappings <- mappings
+	return nil
+}
+
+func (q *querier) queryClientsOfAgent(agent uuid.UUID) ([]mapping, error) {
+	clients, err := q.store.GetTailnetClientsForAgent(q.ctx, agent)
+	if err != nil && !xerrors.Is(err, sql.ErrNoRows) {
+		return nil, err
+	}
+	mappings := make([]mapping, 0, len(clients))
+	for _, client := range clients {
+		node := new(agpl.Node)
+		err := json.Unmarshal(client.Node, node)
+		if err != nil {
+			q.logger.Error(q.ctx, "failed to unmarshal node", slog.Error(err))
+			return nil, backoff.Permanent(err)
+		}
+		mappings = append(mappings, mapping{
+			client:      client.ID,
+			agent:       client.AgentID,
+			coordinator: client.CoordinatorID,
+			updatedAt:   client.UpdatedAt,
+			node:        node,
+		})
+	}
+	return mappings, nil
+}
+
+func (q *querier) queryAgent(agentID uuid.UUID) ([]mapping, error) {
+	agents, err := q.store.GetTailnetAgents(q.ctx, agentID)
+	if err != nil && !xerrors.Is(err, sql.ErrNoRows) {
+		return nil, err
+	}
+	mappings := make([]mapping, 0, len(agents))
+	for _, agent := range agents {
+		node := new(agpl.Node)
+		err := json.Unmarshal(agent.Node, node)
+		if err != nil {
+			q.logger.Error(q.ctx, "failed to unmarshal node", slog.Error(err))
+			return nil, backoff.Permanent(err)
+		}
+		mappings = append(mappings, mapping{
+			agent:       agent.ID,
+			coordinator: agent.CoordinatorID,
+			updatedAt:   agent.UpdatedAt,
+			node:        node,
+		})
+	}
+	return mappings, nil
+}
+
+func (q *querier) subscribe() {
+	eb := backoff.NewExponentialBackOff()
+	eb.MaxElapsedTime = 0 // retry indefinitely
+	eb.MaxInterval = dbMaxBackoff
+	bkoff := backoff.WithContext(eb, q.ctx)
+	var cancelClient context.CancelFunc
+	err := backoff.Retry(func() error {
+		cancelFn, err := q.pubsub.SubscribeWithErr(eventClientUpdate, q.listenClient)
+		if err != nil {
+			q.logger.Warn(q.ctx, "failed to subscribe to client updates", slog.Error(err))
+			return err
+		}
+		cancelClient = cancelFn
+		return nil
+	}, bkoff)
+	if err != nil {
+		// this should only happen if context is canceled
+		return
+	}
+	defer cancelClient()
+	bkoff.Reset()
+
+	var cancelAgent context.CancelFunc
+	err = backoff.Retry(func() error {
+		cancelFn, err := q.pubsub.SubscribeWithErr(eventAgentUpdate, q.listenAgent)
+		if err != nil {
+			q.logger.Warn(q.ctx, "failed to subscribe to agent updates", slog.Error(err))
+			return err
+		}
+		cancelAgent = cancelFn
+		return nil
+	}, bkoff)
+	if err != nil {
+		// this should only happen if context is canceled
+		return
+	}
+	defer cancelAgent()
+
+	// hold subscriptions open until context is canceled
+	<-q.ctx.Done()
+}
+
+func (q *querier) listenClient(_ context.Context, msg []byte, err error) {
+	if xerrors.Is(err, pubsub.ErrDroppedMessages) {
```

Comment:
If `listenClient` expects only `pubsub.ErrDroppedMessages`, shouldn't it be extracted to another wrapping function?

Reply:
I'm afraid I don't understand what you're suggesting.  Can you elaborate?

Reply:
Something like this to ensure type-safety on the method declaration level:

```go

func (q *querier) listenClientRaw(_ context.Context, msg []byte, err error) {
   if err != nil && !xerrors.Is(err, pubsub.ErrDroppedMessages) {
     return
   }
   droppedMessagesErr := err.(pubsub.ErrDroppedMessages)

   q.listenClient(msg, droppedMessagesErr)
}

func (q *querier) listenClient(msg []byte, err pubsub.ErrDroppedMessages) {
...
}
```

Reply:
`listenClient()` is a pubsub listener with error.  The pubsub either sends a message or an error.  If `err` is non-nil, then the `msg` is zero length and not an actual pubsub message.

Right now, the only error that the pubsub can send, and the only one this code handles is `ErrDroppedMessages`.